From: Lorenzo Stoakes <lstoakes@gmail.com>
Date: Wed, 18 Feb 2015 19:24:35 +0000
Subject: [f8cdc7cef553] [PATCH v2] Task 11: Input: wacom - add eudyptula id file to sysfs

This patch adds an id file to the wacom driver's sysfs device directory which,
when read, outputs my Eudyptula ID (f8cdc7cef553) and, when written, raises an
invalid argument error unless the string written to it exactly matches this ID.


[~/linux]$ cat .config | grep -i wacom
# CONFIG_TABLET_SERIAL_WACOM4 is not set
# CONFIG_TOUCHSCREEN_WACOM_W8001 is not set
# CONFIG_TOUCHSCREEN_WACOM_I2C is not set
CONFIG_HID_WACOM=y
[~/linux]$ git diff HEAD~1 | grep -i eudyptula
+/* Eudyptula ID Functionality */
+	/* Eudyptula demands an ID! */
[~/linux]$ make -j8
  HOSTCC  scripts/basic/fixdep
  CHK     include/config/kernel.release
  CHK     include/generated/uapi/linux/version.h
  CHK     include/generated/utsrelease.h
  HOSTCC  scripts/kallsyms
  HOSTCC  scripts/pnmtologo
  HOSTCC  scripts/conmakehash
  HOSTCC  scripts/sortextable
  HOSTCC  scripts/selinux/mdp/mdp
  CC      scripts/mod/empty.o
  HOSTCC  scripts/selinux/genheaders/genheaders

<snip>

  CC      drivers/hid/wacom_wac.o

<snip>

  CC      drivers/hid/wacom_sys.o

<snip>

  LD      drivers/hid/wacom.o

<snip>

  MKPIGGY arch/x86/boot/compressed/piggy.S
  AS      arch/x86/boot/compressed/piggy.o
  LD      arch/x86/boot/compressed/vmlinux
  ZOFFSET arch/x86/boot/zoffset.h
  OBJCOPY arch/x86/boot/vmlinux.bin
  AS      arch/x86/boot/header.o
  LD      arch/x86/boot/setup.elf
  OBJCOPY arch/x86/boot/setup.bin
  BUILD   arch/x86/boot/bzImage
Setup is 15660 bytes (padded to 15872 bytes).
System is 5875 kB
CRC 98927e9
Kernel: arch/x86/boot/bzImage is ready  (#3)
[~/linux]$ cd ~/kerndev
[~/kerndev]$ lsusb | grep -i wacom
Bus 003 Device 006: ID 056a:030e Wacom Co., Ltd Intuos Pen Small (CTL480)
[~/kerndev]$ cat run.sh
#!/bin/bash

KERNEL=/home/lorenzo/linux/arch/x86/boot/bzImage
INITRD=init.img
ROOTFS=rootfs.img
APPEND="root=/dev/sda rw console=ttyS0"
#NETOPTS="-net nic -net bridge,br=br0"
OPTS="-enable-kvm -nographic -m 512M -cpu host -smp 4 $NETOPTS"
DEVS="-usb -device usb-host,hostbus=3,hostaddr=6"

qemu-system-x86_64 $OPTS $DEVS -hda $ROOTFS -initrd $INITRD -kernel $KERNEL -append "$APPEND"
[~/kerndev]$ sudo ./run.sh
[sudo] password for lorenzo:
[    0.000000] Initializing cgroup subsys cpuset
[    0.000000] Initializing cgroup subsys cpu
[    0.000000] Initializing cgroup subsys cpuacct
[    0.000000] Linux version 3.19.0-next-20150212+ (lorenzo@lucifer) (gcc version 4.9.2 20150204 (prerelease) (GCC) ) #4 SMP Thu Feb 12 23:08:26 GMT 2015
[    0.000000] Command line: root=/dev/sda console=ttyS0
[    0.000000] e820: BIOS-provided physical RAM map:

<snip>

  OK  ] Started D-Bus System Message Bus.
[    2.531815] input: Wacom Intuos S Pen as /devices/pci0000:00/0000:00:01.2/usb1/1-1/1-1:1.0/0003:056A:030E.0001/input/input4
[    2.532803] wacom 0003:056A:030E.0001: hidraw0: USB HID v1.10 Device [Wacom Co.,Ltd. Intuos PS] on usb-0000:00:01.2-1/input0
[  OK  ] Reached target Timers.
[    2.542198] input: Wacom Intuos S Pad as /devices/pci0000:00/0000:00:01.2/usb1/1-1/1-1:1.1/0003:056A:030E.0002/input/input7
[    2.543075] wacom 0003:056A:030E.0002: hidraw1: USB HID v1.10 Device [Wacom Co.,Ltd. Intuos PS] on usb-0000:00:01.2-1/input1
<snip>

Arch Linux 3.19.0-next-20150218+ (ttyS0)

localhost login: lorenzo
Password:
[    6.982827] audit: type=1006 audit(1424285923.981:2): pid=2048 uid=0 subj=kernel old-auid=4294967295 auid=1000 old-ses=4294967295 ses=1 res=1
[    6.995629] audit: type=1006 audit(1424285923.994:3): pid=2061 uid=0 subj=kernel old-auid=4294967295 auid=1000 old-ses=4294967295 ses=2 res=1
Last login: Wed Feb 18 18:57:48 on ttyS0
[lorenzo@localhost ~]$ lsusb
Bus 001 Device 002: ID 056a:030e Wacom Co., Ltd Intuos Pen Small (CTL480)
Bus 001 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
[lorenzo@localhost ~]$ cd /sys/devices/pci0000\:00/0000\:00\:01.2/usb1/1-1/1-1\:1.0/0003\:056A\:030E.0001/
[lorenzo@localhost 0003:056A:030E.0001]$ cat id
f8cdc7cef553
[lorenzo@localhost 0003:056A:030E.0001]$ echo foo > id
-bash: id: Permission denied
[lorenzo@localhost 0003:056A:030E.0001]$ echo foo | sudo tee id
[sudo] password for lorenzo:
foo
tee: id: Invalid argument
[lorenzo@localhost 0003:056A:030E.0001]$ echo f8cdc7cef553 | sudo tee id
f8cdc7cef553
[lorenzo@localhost 0003:056A:030E.0001]$

Signed-off-by: Lorenzo Stoakes <lstoakes@gmail.com>
---
 drivers/hid/wacom_sys.c | 44 ++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 44 insertions(+)

diff --git a/drivers/hid/wacom_sys.c b/drivers/hid/wacom_sys.c
index f0568a7..8763028 100644
--- a/drivers/hid/wacom_sys.c
+++ b/drivers/hid/wacom_sys.c
@@ -27,6 +27,39 @@
 #define DEV_ATTR_RW_PERM (S_IRUGO | S_IWUSR | S_IWGRP)
 #define DEV_ATTR_WO_PERM (S_IWUSR | S_IWGRP)

+/* Eudyptula ID Functionality */
+
+static const char hello_id[] = "f8cdc7cef553";
+static const size_t hello_id_len = 12;
+
+static ssize_t
+id_show(struct device *dev, struct device_attribute *attr, char *buf)
+{
+	return sprintf(buf, "%s\n", hello_id);
+}
+
+static ssize_t
+id_store(struct device *dev, struct device_attribute *attr, const char *buf,
+	size_t count)
+{
+	if (count < hello_id_len || strncmp(buf, hello_id, hello_id_len))
+		return -EINVAL;
+
+	return count;
+}
+
+static DEVICE_ATTR(id, 0644, id_show, id_store);
+
+static struct attribute *id_attrs[] = {
+	&dev_attr_id.attr,
+	NULL
+};
+
+static struct attribute_group id_attr_group = {
+	.attrs = id_attrs
+};
+
+
 static int wacom_get_report(struct hid_device *hdev, u8 type, u8 *buf,
 			    size_t size, unsigned int retries)
 {
@@ -1518,6 +1551,15 @@ static int wacom_probe(struct hid_device *hdev,
 			wacom_wac->shared->touch_input = wacom_wac->input;
 	}

+	/* Eudyptula demands an ID! */
+	error = sysfs_create_group(&wacom->hdev->dev.kobj, &id_attr_group);
+
+	if (error) {
+		hid_err(wacom->hdev,
+			"cannot create sysfs group err: %d\n", error);
+		return error;
+	}
+
 	return 0;

 fail_hw_start:
@@ -1552,6 +1594,8 @@ static void wacom_remove(struct hid_device *hdev)
 	wacom_destroy_battery(wacom);
 	wacom_remove_shared_data(&wacom->wacom_wac);

+	sysfs_remove_group(&wacom->hdev->dev.kobj, &id_attr_group);
+
 	hid_set_drvdata(hdev, NULL);
 	kfree(wacom);
 }
--
2.3.0
